src/qc.rs

```diff
@@ -142,3 +142,123 @@ macro_rules! arbitrary_large {
+//            if a.0 == 0 {
+//                return None
+//            }
+//
+//            // Return the result via `Some` if the test can run
+//            let mut other_result = 0;
+//            let result = f(a.0, &mut other_result);
+//            Some((result, other_result))
+//        }
+//    }
+//
+// If anything returns `None` then the test is discarded, otherwise the two
+// results are compared for equality and the test fails if this equality check
+// fails.
+macro_rules! check {
+    ($(
+        fn $name:ident($f:ident: extern fn($($farg:ty),*) -> $fret:ty,
+                       $($arg:ident: $t:ty),*)
+                       -> Option<$ret:ty>
+        {
+            $($code:tt)*
+        }
+    )*) => (
+        $(
+            fn $name($f: extern fn($($farg),*) -> $fret,
+                     $($arg: $t),*) -> Option<$ret> {
+                $($code)*
+            }
+        )*
+
+        mod _compiler_rt {
+            use qc::*;
+            use std::mem;
+            use quickcheck::TestResult;
+
+            $(
+                #[test]
+                fn $name() {
+                    fn my_check($($arg:$t),*) -> TestResult {
+                        let my_answer = super::$name(super::super::$name,
+                                                     $($arg),*);
+                        let compiler_rt_fn = ::compiler_rt::get(stringify!($name));
+                        unsafe {
+                            let compiler_rt_answer =
+                                super::$name(mem::transmute(compiler_rt_fn),
+                                             $($arg),*);
+                            match (my_answer, compiler_rt_answer) {
+                                (None, _) | (_, None) => TestResult::discard(),
+                                (Some(a), Some(b)) => {
+                                    TestResult::from_bool(a == b)
+                                }
+                            }
+                        }
+                    }
+
+                    ::quickcheck::quickcheck(my_check as fn($($t),*) -> TestResult)
+                }
+            )*
+        }
+
+        mod _gcc_s {
+            use qc::*;
+            use std::mem;
+            use quickcheck::TestResult;
+
+            $(
+                #[test]
+                fn $name() {
+                    fn my_check($($arg:$t),*) -> TestResult {
+                        let my_answer = super::$name(super::super::$name,
+                                                     $($arg),*);
+                        let gcc_s_fn = match ::gcc_s::get(stringify!($name)) {
+                            Some(f) => f,
+                            // If it's not in libgcc, or we couldn't find
+                            // libgcc, then just ignore this. We should have
+                            // tests through compiler-rt in any case
+                            None => return TestResult::passed(),
```

Comment:
could the logic be changed to avoid calling quickcheck::quickcheck at all if the gcc_s symbol is not available?